finality-aleph/src/data_io.rs

```diff
@@ -233,46 +231,23 @@ where
 }
 
 #[derive(Clone)]
-pub(crate) struct DataIO<B: BlockT> {
-    pub(crate) best_chain: Arc<Mutex<AlephDataFor<B>>>,
+pub(crate) struct DataIO<B: BlockT, SC: SelectChain<B>> {
+    pub(crate) select_chain: SC,
     pub(crate) ordered_batch_tx: mpsc::UnboundedSender<OrderedBatch<AlephDataFor<B>>>,
     pub(crate) metrics: Option<Metrics<B::Header>>,
 }
 
-pub(crate) async fn refresh_best_chain<B: BlockT, SC: SelectChain<B>>(
-    select_chain: SC,
-    best_chain: Arc<Mutex<AlephDataFor<B>>>,
-    mut exit: oneshot::Receiver<()>,
-) {
-    loop {
-        let delay = futures_timer::Delay::new(Duration::from_millis(REFRESH_INTERVAL));
-        tokio::select! {
-            _ = delay => {
-                let new_best_header = select_chain
-                    .best_chain()
-                    .await
-                    .expect("No best chain");
-                *best_chain.lock() = AlephData::new(new_best_header.hash(), *new_best_header.number());
-            }
-            _ = &mut exit => {
-                debug!(target: "afa", "Task for refreshing best chain received exit signal. Terminating.");
-                return;
-            }
-        }
-    }
-}
-
-impl<B: BlockT> aleph_bft::DataIO<AlephDataFor<B>> for DataIO<B> {
+impl<B: BlockT, SC: SelectChain<B>> aleph_bft::DataIO<AlephDataFor<B>> for DataIO<B, SC> {
     type Error = Error;
 
     fn get_data(&self) -> AlephDataFor<B> {
-        let best = *self.best_chain.lock();
+        let new_best_header = self.select_chain.best_chain().expect("No best chain");
```

Comment:
Just to make sure I understand correctly – `best_chain` is not async in 3.0.0, so we don't need all the mechanism we had for refreshing it before?

Reply:
Precisely.